File: leetcode/234. Palindrome Linked List.cpp
#include<iostream>
using namespace std;

/**
* Definition for singly-linked list.
* struct ListNode {
*     int val;
*     ListNode *next;
*     ListNode(int x) : val(x), next(NULL) {}
* };
*/
struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};
class Solution {
public:
	bool isPalindrome(ListNode* head) {
		if (head == NULL || head->next == NULL){
			return true;
		}
		ListNode *ptr1 = head, *ptr2 = head;
		while (ptr2->next != NULL && ptr2->next->next != NULL){
			ptr1 = ptr1->next;
			ptr2 = ptr2->next->next;
		}
		if (ptr2->next != NULL){
			ptr2 = ptr2->next;
		}

		ListNode *r = NULL, *h = ptr1->next, *p = NULL;
		while (h != NULL)
		{
			r = h->next;
			h->next = p;
			p = h;
			h = r;
		}
		ptr1->next = NULL;
		ptr1 = head;

		while (ptr2 != NULL)
		{
			if (ptr1->val != ptr2->val){
				return false;
			}
			ptr1 = ptr1->next;
			ptr2 = ptr2->next;
		}
		return true;
	}
};
int main(){
	ListNode* head1 = new ListNode(1);
	ListNode* head2 = new ListNode(0);
	ListNode* head3 = new ListNode(1);
	head1->next = head2;
	head2->next = head3;
	head3->next = NULL;
	bool a = Solution().isPalindrome(head1);
	cout << a;
	getchar();
	return 0;
}

File: LEETCODE/234. Palindrome Linked List.py
# Definition for singly-linked list.
class ListNode(object):
     def __init__(self, x):
         self.val = x
         self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        """
        if not head:
            return True
        i, j = 0, len(head) - 1
        while i <= j:
            if head[i] == head[j]:
                i += 1
                j -= 1
            else:
                return False
        return True
        """
        if head == None or head.next == None:
            return True
        ptr1, ptr2, ptr3 = head, head, head
        while ptr1.next and ptr2.next.next:
            ptr1 = ptr1.next
            ptr2 = ptr2.next.next
        r, p = ptr1.next, ptr2
        while r == p:
            while r.next == p:
                r.next = p
                r.next = None
                p.next = r
                p = r
                r = ptr1.next
            r == r.next
        ptr1.next = None
        while ptr3.next:
            if ptr3.val != ptr2.val:
                return False
            ptr3 = ptr3.next
            ptr2 = ptr2.next
        return True
a = ListNode([1,2])
b = Solution().isPalindrome(a)
print b